### PR TITLE
Talos

### LINKED ISSUE #888: Create psammead bumper bot (strictly for psammead packages)
**Is your feature request related to a problem? Please describe.**
We are falling behind on psammead bumps, we should create a bot to bump dependancies for us. We tried dependabot but it opened 100PRs with single bumps per PR which isn't ideal

**Describe the solution you'd like**
When a psammead PR is merged & published
 -> For each newly published package, bump all dependants (patch, minor, major) in psammead
 -> Run `npm run updateSnapshots`
 ->  -> Make a single draft PR with a summary of the changes.

The changelog and version bumping need to be an automated part of this solution. This would probably best be done as a separate PR that precedes this primary implementation.

**Testing notes**
No testing required

**Additional context**
Please write in JS or bash. Not enough devs here write python so its unmaintainable.
Should be public and OSS, ideally in this repo.

### PR DESCRIPTION
**Pull request by dr3**
Resolves #888

**Overall change:** Bumper bot for psammead

**Code changes:**

- Create talos

This doesnt do some stuff, like bump the base level dependancy. this wasnt done as this PR was done before we moved all dev deps to the base level, this will be added soon as an iteration. :)

---

- [ ] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval

### CODE REVIEW

File: scripts/talos/commitChanges/index.js
Comment (jamesbhobbs) [inline scripts/talos/commitChanges/index.js:4]: Wouldn't `git add packages` miss the main package.json? Which should also be bumped with the package version numbers

Comment (dr3) [reply scripts/talos/commitChanges/index.js:4]: Yeah the main body of this PR was done before we moved all the psammead package dev dept to root. I was going to raise and issue to bump the base one after this is merged, but I can do it now if you think its high enough importance?

**Review by jamesbhobbs — COMMENTED** [review]
From a skim read... Nice approach and looks thoroughly tested.

File: Jenkinsfile
Comment (chris-hinds) [inline Jenkinsfile:120]: Why do we need to do this? and does need to be 777?

File: scripts/talos/checkoutBranch/index.test.js
Comment (chris-hinds) [inline scripts/talos/checkoutBranch/index.test.js]:
```suggestion
  it('should not catch its own errors', async () => {
```

File: scripts/talos/createPullRequest/index.js
Comment (chris-hinds) [inline scripts/talos/createPullRequest/index.js:10]: Sounds silly but this could end up being a really long title in some bumps.

File: scripts/talos/createPullRequest/index.js
Comment (chris-hinds) [inline scripts/talos/createPullRequest/index.js]: Do we want these to open as draft PR's? or is this a temporary state?

File: scripts/talos/createPullRequest/index.test.js
Comment (chris-hinds) [inline scripts/talos/createPullRequest/index.test.js]:
```suggestion
  it('should not catch its own errors', async () => {
```

File: scripts/talos/getChangedPackages/index.test.js
Comment (chris-hinds) [inline scripts/talos/getChangedPackages/index.test.js]:
```suggestion
  it('should return an empty array if the file is empty', async () => {
```

File: scripts/talos/getChangedPackages/index.test.js
Comment (chris-hinds) [inline scripts/talos/getChangedPackages/index.test.js]:
```suggestion
  it('should return an empty array if file doesn't exist', async () => {
```

File: scripts/talos/index.test.js
Comment (chris-hinds) [inline scripts/talos/index.test.js]: this could be a little more descriptive :P

**Review by chris-hinds — CHANGES_REQUESTED** [review]
@dr3 this is really nice! Just left a few comments/suggestions.

File: Jenkinsfile
Comment (dr3) [reply Jenkinsfile:120]: Needed because there were issues where the `.git` folder was changed (via one of the git commands), which meant when the pipeline next ran, when jenkins tried to check out the branch, it couldnt as it didnt have write access. 777 is a 'give everything back to jenkins' permision level :P

File: scripts/talos/createPullRequest/index.js
Comment (dr3) [reply scripts/talos/createPullRequest/index.js:10]: Yeah, github will limit them and put the rest in the Pr body, fine for now IMO, can iterate when we improve the PR body

File: scripts/talos/createPullRequest/index.js
Comment (dr3) [reply scripts/talos/createPullRequest/index.js]: ermm I actually dont think `draft: true` even works :P ile remove it

File: scripts/talos/index.test.js
Comment (dr3) [inline scripts/talos/index.test.js]:
```suggestion
  it('should run expected commands to create PR', async () => {
```

File: scripts/talos/createPullRequest/index.js
Comment (dr3) [inline scripts/talos/createPullRequest/index.js]:
```suggestion
```

File: scripts/talos/createPullRequest/index.test.js
Comment (dr3) [inline scripts/talos/createPullRequest/index.test.js]:
```suggestion
```

File: Jenkinsfile
Comment (chris-hinds) [reply Jenkinsfile:120]: Does this not work with `755` doing a chmod to 777 is never a good thing.

Comment (dr3) [reply Jenkinsfile:120]: I dont think `755` will work as that doesnt allow public write access, which i think we need to allow jenkins to checkout and change the .git folder

**Review by chris-hinds — APPROVED** [review]

**Review by thekp — APPROVED** [review]

**Comment by dr3** [comment]
Didnt need testing, will fix forward

File: Jenkinsfile
Comment (jamesdonoh) [reply Jenkinsfile:120]: This is OK to get going but we need a permanent solution that doesn't grant global write permission. There must be another way to allow subsequent runs to overwrite existing files, e.g. have we checked the owner/group of these files? Please can you create an issue to revisit?

File: scripts/talos/index.test.js
Comment (jamesdonoh) [inline scripts/talos/index.test.js:108]: Could specify the non-zero error code here as this is important behaviour for the pipeline.

File: scripts/talos/getBranchName/index.js
Comment (jamesdonoh) [inline scripts/talos/getBranchName/index.js:9]: How confident are we that two instances of `talos` can never be running at the same second and therefore both interact on the same branch?

**Pull request merged**
